### PR TITLE
[SPARK-22143][SQL] Fix memory leak in OffHeapColumnVector

### PR DESCRIPTION
**Pull request by hvanhovell**
## What changes were proposed in this pull request?
`WriteableColumnVector` does not close its child column vectors. This can create memory leaks for `OffHeapColumnVector` where we do not clean up the memory allocated by a vectors children. This can be especially bad for string columns (which uses a child byte column vector).

## How was this patch tested?
I have updated the existing tests to always use both on-heap and off-heap vectors. Testing and diagnoses was done locally.

### CODE REVIEW

**Comment by hvanhovell** [comment]
cc @ala @michal-databricks @ueshin

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java [hunk 1]
```diff
@@ -85,6 +85,7 @@ public long nullsNativeAddress() {
 
   @Override
   public void close() {
+    super.close();
     Platform.freeMemory(nulls);
```
Comment (kiszk) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java]: This comment is not related to this fix directly. However, is it better to check whether each field is not `0` before calling `Platform.freeMemory()`? For example, `data` or `lengthData/offsetData may be `0`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala [hunk 2]
```diff
@@ -25,19 +25,25 @@ import org.apache.spark.sql.types._
...
 
-  override def afterEach(): Unit = {
-    testVector.close()
+  private def testVectors(
+      name: String,
+      size: Int,
+      dt: DataType)(
+      block: WritableColumnVector => Unit): Unit = {
+    test(name) {
+      val c1 = new OnHeapColumnVector(size, dt)
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala]: What's this `c1` for?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 3]
```diff
@@ -718,62 +705,69 @@ class ColumnarBatchSuite extends SparkFunSuite {
       column.putArray(0, 0, array.length)
       assert(ColumnVectorUtils.toPrimitiveJavaArray(column.getArray(0)).asInstanceOf[Array[Int]]
         === array)
-    }}
   }
 
   test("toArray for primitive types") {
-    // (MemoryMode.ON_HEAP :: MemoryMode.OFF_HEAP :: Nil).foreach { memMode => {
-    (MemoryMode.ON_HEAP :: Nil).foreach { memMode => {
+    (MemoryMode.ON_HEAP :: MemoryMode.OFF_HEAP :: Nil).foreach { memMode => {
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: We can remove a pair of brace like `.foreach { memMode => ...`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 4]
```diff
@@ -718,62 +705,69 @@ class ColumnarBatchSuite extends SparkFunSuite {
...
+    // (MemoryMode.ON_HEAP :: Nil).foreach { memMode => {
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: Can we remove this line?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java (hunk 1 above)
Comment (hvanhovell) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/OffHeapColumnVector.java]: Yeah, it is a bit weird. The doc in the JDK8's `Unsafe. freeMemory()` states the following: `The address passed to this method may be null, in which case no action is taken.` See: http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/misc/Unsafe.java

**Review by ueshin — COMMENTED** [review]
LGTM except for some minor comments.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala (hunk 2 above)
Comment (ala) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala]: What is `c1` for?

**Review by michal-databricks — APPROVED** [review]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 5]
```diff
@@ -38,31 +38,45 @@ import org.apache.spark.unsafe.types.CalendarInterval
...
+  private def testVector(
+      name: String,
+      size: Int,
+      dt: DataType,
+      modes: MemoryMode*)(
+      block: (WritableColumnVector, MemoryMode) => Unit): Unit = {
+    test(name) {
+      modes.foreach { mode =>
+        val vector = allocate(size, dt, mode)
+        try block(vector, mode) finally {
```
Comment (ala) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: The way you use `{}` here is weird.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 6]
```diff
@@ -422,17 +420,14 @@ class ColumnarBatchSuite extends SparkFunSuite {
           assert(v._1 == Platform.getLong(null, addr + 8 * v._2))
         }
       }
-    }}
   }
 
-  test("Float APIs") {
-    (MemoryMode.ON_HEAP :: MemoryMode.OFF_HEAP :: Nil).foreach { memMode => {
+  testVector("Float Apis", 1024, FloatType, MemoryMode.ON_HEAP, MemoryMode.OFF_HEAP) {
```
Comment (ala) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: I think changing `Apis` to `APIs` would be better than the other way around.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 7]
```diff
@@ -829,13 +823,15 @@ class ColumnarBatchSuite extends SparkFunSuite {
...
-  test("Nest Struct in Array.") {
-    (MemoryMode.ON_HEAP :: MemoryMode.OFF_HEAP :: Nil).foreach { memMode =>
-      val schema = new StructType().add("int", IntegerType).add("long", LongType)
-      val column = allocate(10, new ArrayType(schema, true), memMode)
+  private val structType: StructType = new StructType().add("i", IntegerType).add("l", LongType)
+
+  testVector(
+    "Nest Struct in Array",
+    10,
+    new ArrayType(new StructType().add("int", IntegerType).add("long", LongType), true),
```
Comment (ala) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: Did you mean `new ArrayType(structType)`?

**Review by ala — APPROVED** [review]
LGTM. Just some nitpicks.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala [hunk 8]
```diff
@@ -181,9 +177,8 @@ class ColumnVectorSuite extends SparkFunSuite with BeforeAndAfterEach {
     assert(array.get(3, arrayType).asInstanceOf[ArrayData].toIntArray() === Array(3, 4, 5))
   }
 
-  test("struct") {
-    val schema = new StructType().add("int", IntegerType).add("double", DoubleType)
-    testVector = allocate(10, schema)
+  val structType: StructType = new StructType().add("int", IntegerType).add("double", DoubleType)
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala]: nit: Do we need `: StructType`?

Comment (hvanhovell) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnVectorSuite.scala]: It is a class level field now, so scalastyle wants me type it...

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 9]
```diff
@@ -38,31 +38,45 @@ import org.apache.spark.unsafe.types.CalendarInterval
...
   }
 
-  test("Null Apis") {
-    (MemoryMode.ON_HEAP :: MemoryMode.OFF_HEAP :: Nil).foreach { memMode => {
-      val reference = mutable.ArrayBuffer.empty[Boolean]
+  private def testVector(
+      name: String,
+      size: Int,
+      dt: DataType,
+      modes: MemoryMode*)(
```
Comment (kiszk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: Can we do `modes: Seq[MemoryMode] = Seq(MemoryMode.ON_HEAP, MemoryMode.ON_HEAP))(`? Then, we can avoid repeat `testVector(..., MemoryMode.ON_HEAP, MemoryMode.OFF_HEAP)` at each suite.

**Comment by hvanhovell** [comment]
retest this please

**Comment by hvanhovell** [comment]
Merging to master.
